src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CommandMessageBinaryEncoder.cs

```diff
@@ -296,10 +296,7 @@ private void WriteType1Section(BsonBinaryWriter writer, Type1CommandMessageSecti
             stream.WriteCString(section.Identifier);
 
             var batch = section.Documents;
-            var maxDocumentSize = 
-                IsEncryptionConfigured && MaxDocumentSize.HasValue
-                    ? MaxDocumentSize.Value 
-                    : section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
+            var maxDocumentSize = section.MaxDocumentSize ?? writer.Settings.MaxDocumentSize;
```

Comment:
Good... `maxDocumentSize` is now independent of whether FLE is in use or not.

Reply:
What we need now is to use a smaller `maxMessageSize` when FLE is in use. Replace line 292 above with:

    var maxMessageSize = MaxMessageSize;
    if (IsEncryptionConfigured)
    {
        var maxBatchSize = 2 * 1024 * 1024; // 2 MiB
        var maxMessageEndPosition = stream.Position + maxBatchSize;
        maxMessageSize = (int)(maxMessageEndPosition - messageStartPosition);
    }

The logic here is:

* we want to enforce a maximum *batch* size of 2MiB
* the easiest way to enforce that is to compute what the message size would be when the batch is full

Reply:
The way I wrote it emphasizes that we are *overriding* the `maxMessageSize` when encryption is configured, but the way you wrote it works also.

Reply:
I wanted to highlight that `maxMessageSize` can be calculated in two different ways. And these ways are mutually exclusive.

Reply:
OK. You wanted to highlight the *opposite* of what I wanted to highlight!

That's OK too.